### ISSUE TITLE
refactoring the MDD provider to separate concerns

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#735]
**Reported by @manuelbernhardt on 20 Apr 2010 11:04 UTC**
the new MDD provider should be more generic and less parser-dependent, so it can be more easily extended for other parse sources.

here is a design proposal:

== org.makumba.providers.mdd

DataDefinitionImpl				implementation of DataDefinition
FieldDefinitionImpl				implementation of FieldDefinition
MDDProvider						concrete subclass of the DataDefinitionProvider, uses an
								AbstractDataDefinitionFactory in order to do its stuff
AbstractDataDefinitionFactory		factory for DataDefinition and FieldDefinition. provides methods for
								building DataDefinition-s and FieldDefinition-s, used by the
								DataDefinitionProvider subclass

== org.makumba.providers.mdd.parser

MDDDataDefinitionFactory			concrete factory for MDDs based on the ANTLR parser.
								replaces MDDFactory

all the classes related to ANTLR parsing. only the MDDDataDefinitionFactory knows about them.

(future: org.makumba.providers.mdd.XMLparser, org.mak.providers.mdd.tableparser, ...)

== org.makumba.providers.mdd.parser.validation

nodes that hold data about the validation rules

== org.makumba.providers.validation

implementation of the validation rules

this design allows it to get DataDefinitions from various sources (e.g. XML) and use the MDD interfaces implementations with any kind of data.

the difficult part about MDDs is to build them, i.e. to link the MDD to other MDDs and to subfields at build-time. this process is especially tedious as there is an implicit contract with the TableManager that expects naming conflicts (for MDDs that have fields with the same name as the MDD, etc.) to be solved in a specific fashion. Thus, the implementations of the AbstractDataDefinitionFactory should deal with this MDD-building issue. further they need to be able to work with the caching mechanism so that they don't try to re-create themselves at build time.

in practice this means that DataDefinitionImpl#build() as well as some constructors of the DataDefinitionImpl and the FieldDefinitionImpl need to be moved to the MDDDataDefinitionFactory, which replaces MDDFactory.

further the duplicated code in the validation rules gets removed. probably there will be a need to make a super-class for the VDDs as the new implementations make a difference between single-field and multi-field VDDs thus the mechanism there is actually different.

see org.makumba.providers.datadefinition.mdd.ValidationRuleNode:

    // FIXME remove this once we totally migrate to the new VDs
    public FieldDefinition getFieldDefinition() {
        // TODO Auto-generated method stub
        return null;
    }

==> the problem with the interface method getFieldDefinition() is that it does not apply for multi-field validation rules. thus perhaps making a new interface for those makes sense.

yet, that would also mean adapting the implementation of the old VDDs, not sure how much effort that means.

Migrated-From: http://trac.makumba.org/ticket/1190

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Apr 2010 11:35 UTC**
i think there are two concerns and there should be "providers" for both of them.

1) MDD data (MDDs, fields, functions, rules) storing. this is not a difficult matter, and one provider is enough. once the tests pass we can throw away the rest. but this provider should be separated from (2).

i don't think we need to (over)design any big contract for this  besides the contract that already exists (DataDefinition, FieldDefinition, ...)

for tests, what I think is needed is a comparator, to compare the data in 2 DataDefinitions, FieldDefinitions, etc. this would allow for corpus-based work. for function definitions there is already an AST comparator. we need comparisons for FieldDefinitions and validation definitions. due to sensitivities related to order of fields at lower makumba levels, that should also be checked here.

2) MDD parsing, i.e. obtaining the MDD data from various sources. 

for the current .mdd file source, i would even go as far as, instead of refactoring the existing parsers, do a new, one-pass parser that combines the MDD, MQL and maybe regexp grammars (see Filip's experience). if there is an MDD comparator, we can go straight for this, as we can confidently check that the new parser will return the same results.

the reason i propose this is that what i learned from past antlr experiences is that it's good to have a grammar that matches everything, and doesn't try to ignore parts (like function bodies) to parse them separately. being able to call a grammar from another would be great.

also i suspect quite a lot that since the mdd grammar is so simple, a one-pass parser is enough, and would allow us to keep things simple. certainly for functions only one pass is needed at MDD parse time.

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 20 Apr 2010 12:58 UTC**
There are two other concerns which have to be considered: the MDD functionality (DataDefinitions and FieldDefinitions are not only data) and the MDD assembling / building.

Especially the latter part (assembling) is complicated due to the implicit contract with the database layer. This is why I propose to make this concern explicit by making a DataDefinitionFactory which aims at addressing this point.

Also, this concern is heavily dependent on the nature of the parser at hand: the RecordParser does a line-by-line parsing followed by some checking and data manipulation, which is very different of a ANTLR parser which follows rules. This is where most of the "magic" behind MDDs happens. Thus my proposal to have an abstract factory (that can be used later on for other sources) and a concrete implementation as part of the antlr parser.


I am not sure if making separate top-level makumba providers for the parsing, assembly and functionality / data holding makes so much sense. I mean the current interface (DataDefinitionProvider) is used everywhere, and the fuzzy contract dealt with by the assembly would make the design of such a provider hard to say the least.

Thus I would propose (for this bug entry) to focus on making a cleaner separation between
- mdd data & functionality (DataDefinitionImpl and FieldDefinitionImpl, ValidationRules and Functions)
- antlr parser
- MDD assembly (builder)

At the moment the separation between the (DDImpl, FDImpl) and antlr parser is not that clear as well as validation rule functionality & validation rule data.



Two other issues are:
1) comparison with the old mdd parser
2) antlr parser simplification

However these are not what I wrote this bug entry for, so I propose not to discuss them here.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Apr 2010 21:48 UTC**
i make some notes here because i think that this MDD business is getting more complicated than it should be.

what do you mean by assembly? is it at field level or at mdd level?

if the parser makes FieldDefinitions (or just the data for them, but in a class that's parser-independent) and the storage stores them in the right order, is that good separation?

i am struck by the number of methods to make/get field definitions in DataDefinitionProvider. should we not begin by simplifying that, to ease the job of MDD parsers?

(In reply to comment #2)
> There are two other concerns which have to be considered: the MDD functionality
> (DataDefinitions and FieldDefinitions are not only data) and the MDD assembling
> / building.

ok concern is maybe not the right word. what i mean is

1) for storing MDD data and providing functionality about it, we don't need alternatives. it's not a difficult problem, code is already quite clean, there's not much variation.

2) for parsing we need alternative providers because
2.1) we want to be able to easily change the grammar while not affecting old MDDs
2.2) we want to extract MDD data from various types of sources

the important thing here is to be able to vary these two (especially the second) independently, and communicate via an interface (FieldDefinition and DataDefinition are maybe good enough or could be made good enough? or just some FieldDefinitionData as mentioned above?).

and one more concern: _finding_ the MDD file. now we only support 2 classpath based alternatives, but we could support more. 

> Especially the latter part (assembling) is complicated due to the implicit
> contract with the database layer. This is why I propose to make this concern
> explicit by making a DataDefinitionFactory which aims at addressing this point.

well i think _that_ contract must be made explicit, first, before doing anything else. but i think that contract simply counts on some field order. 

- primary key
- TS_modify
- TS_create
other fields.

plus special versions of this in special tables, like sets etc. anything else that the table needs?

> Also, this concern is heavily dependent on the nature of the parser at hand:
> the RecordParser does a line-by-line parsing followed by some checking and data
> manipulation, which is very different of a ANTLR parser which follows rule
> This is where most of the "magic" behind MDDs happens. Thus my proposal to have

magic? it's about reading a field name, a type and some modifiers. really, let us not over-complicate things. maybe i've done a mess in the first parser, and the handler refactoring didn't help to make things better, but still.

> an abstract factory (that can be used later on for other sources) and a
> concrete implementation as part of the antlr parser.

as long as this is just for parsing concern, that's ok. but should we not wait with factories until we know more about this?

like "MDD parser is some code that takes text (obtained by the MDD finder) and makes FieldDefinitionData which it then sends to the storage, along with the name of the MDD that it's part of. it can send fields one by one or in batches". then it's the responsibility of the storage to assemble, etc. 

so 
interface MDDParser{
   void parse(Reader r, MDDStorage s);
}

interface MDDStorage{
   void addField(String mddName, FieldDefinitionData d);
   void addRule(String mddName, RuleData r);
   void addFunction(String mddName, FunctionData f);
}

then it's up to the storage to order the fields, check whether they are ok, etc. all the *Data classes can contain file location information (file, line, column) to be able to format friendly errors. that location data can be dropped once the MDD becomes validated.

> I am not sure if making separate top-level makumba providers for the parsing,
> assembly and functionality / data holding makes so much sense. I mean the

not top level. but internally, depending on file extension, grammar version, we will need to choose a parser. grammar version can be indicated with a comment in the file, like XML does.

actually at the top level it could be as simple as not even showing many providers. simply a static method to get an mdd, and maybe a more sophisticated method where you can indicate parser, mdd finder. the default such things should be configurable in configuration... so there would be three ways to choose the parser:
- the file itself
- through an API
- through configuration

> current interface (DataDefinitionProvider) is used everywhere, and the fuzzy
> contract dealt with by the assembly would make the design of such a provider
> hard to say the least.

> Thus I would propose (for this bug entry) to focus on making a cleaner
> separation between
> - mdd data & functionality (DataDefinitionImpl and FieldDefinitionImpl,
> ValidationRules and Functions)
> - antlr parser
> - MDD assembly (builder)

i don't see the assembler difficulty... is that about making the field data, or about storing field data in a list to make the MDD data?

> At the moment the separation between the (DDImpl, FDImpl) and antlr parser is
> not that clear as well as validation rule functionality & validation rule data.

to what i know Rudi already separated function information from any other MDD data (from either provider) and he was planning to do so for the valuidation stuff too (not sure)

/* also the function AST is no longer part of MDD data. only raw function text (without this-adding) is needed 
*/

> Two other issues are:
> 1) comparison with the old mdd parser
> 2) antlr parser simplification
> 
> However these are not what I wrote this bug entry for, so I propose not to
> discuss them here.

well I think the comparator is relevant here, I mean when there is a corpus with hundreds of MDDs (I made one for Filip today) it would be a pity not to use it to compare the results of the new code with the old (or the two old ones)... I think FieldDefinition can be used for that directly?

in any case i don't think a bugzilla entry for MDD comparators makes sense; we should simply make one :) or if we deem that it's not needed for refactoring, there's no point to make one :)

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 21 Apr 2010 07:26 UTC**
> > Especially the latter part (assembling) is complicated due to the implicit
> > contract with the database layer. This is why I propose to make this concern
> > explicit by making a DataDefinitionFactory which aims at addressing this point.
> 
> well i think _that_ contract must be made explicit, first, before doing
> anything else. but i think that contract simply counts on some field order. 

the simple fields (timestamps, ptrIndex, etc.) are not a problem. the tricky part starts with subfields and especially sets, their pointer to the parent and the foreign. there is a mechanism to add "_" to the name of these fields, and to avoid conflicts with fields that bear the same name than the parent MDD, or the subfield, or the foreign.

e.g. you can have

best.internal.hrtool.Team

team = set general.Team

then, there will be a subtable MDD which contains the mapping between both tables and that has following fields:

team: ptrIndex of the MDD subtable
team_: ptrRel to the general.Team
team__: ptrRel to the best.internal.hrtool.Team

There is the similar kind of problem in other cases: MDDs that have fields having the same name than the MDD and pointing to themselves, and some others, I can't remember all of them now.

If the order is not identical to the one generated by the old parser, you can't use the provider on existing databases.

See some of the collision avoidance mechanisms here:

http://fisheye3.atlassian.com/browse/makumba/trunk/makumba/classes/org/makumba/providers/datadefinition/mdd/DataDefinitionImpl.java?r=4822#l282

Note that my aim when writing the new parser was to make this stuff explicit and duly documented, i.e. when I come back one year later I can still understand why I wrote a specific part of the code. I can't say the same of the RecordParser, that's why I will not touch that code anymore - it just takes way too much time to understand what it does.


Besides of the name conflicting thing, there is also a chicken-and-egg problem, see e.g.

http://fisheye3.atlassian.com/browse/makumba/trunk/makumba/classes/org/makumba/providers/datadefinition/mdd/DataDefinitionImpl.java?r=4822#l210

Finally there is also the issue that when you build an MDD, you cannot build it at once and keep on adding fields because eventually the MDD will point to itself. In that case, the NamedResources will complain that the resource tries to re-create itself, so the building needs to be dealt with in a separate step.

For these reasons, and given the sheer amount of code that only deals with assembling the fields and making the sub-tables for sets etc. in DataDefinitionImpl, I want to move this code into a specific class. In the same time, that would remove the knowledge of ASTs from DataDefinitionImpl and FieldDefinitionImpl.


> interface MDDStorage{
>    void addField(String mddName, FieldDefinitionData d);
>    void addRule(String mddName, RuleData r);
>    void addFunction(String mddName, FunctionData f);
> }


given what I said above, that is not enough. I actually have no idea how you would model it in an interface...maybe

void addPointerRelToParentBeforePointerRelToForeignAndAvoidNameCollisions()

will do.

Seriously, the problem is not _that_ simple.

-----
about the refactoring: my purpose with this entry was to make the new provider somewhat cleaner, given the concerns rudi raised in some mail (modularity with XML, ...), i.e. make the DataDefinitionImpl and FieldDefinitionImpl independent of AST, but _not_ to re-code the whole thing, introduce a MDDStorage, etc., i.e. change the whole design of the provider.

so I think there are two stages:

1) simple refactoring to make DataDefinitionImpl and FieldDefinitionImpl separate from AST, by introducing a factory that takes care of assembling MDDs, subfields etc.
2) complex refactoring, that involves much more work.

I am ok with doing level 1) but not more.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 21 Apr 2010 09:12 UTC**
(In reply to comment #4)
> the foreign. there is a mechanism to add "_" to the name of these fields, and
> to avoid conflicts with fields that bear the same name than the parent MDD, or
> the subfield, or the foreign.

yes. the storage should take care of that. the parser simply spits fields to it. if the parser spits 2 fields with the same name in the same mdd, it's a programmer error. if the parser spits a field with the same name as an internally generated field, we add an underscore, or more, until we get a unique name.

> best.internal.hrtool.Team
> team = set general.Team
> 
> team: ptrIndex of the MDD subtable
> team_: ptrRel to the general.Team
> team__: ptrRel to the best.internal.hrtool.Team

both of these fields are internally generated so this is no problem. the storage can manage that.

my point is: assembly is not dependent on parsing.

> There is the similar kind of problem in other cases: MDDs that have fields
> having the same name than the MDD and pointing to themselves, and some others,
> I can't remember all of them now.

this can all be managed independently of parsing.

> 
> If the order is not identical to the one generated by the old parser, you can't
> use the provider on existing databases.

order can be managed by the storage, independently of parsing.

> Note that my aim when writing the new parser was to make this stuff explicit
> and duly documented, i.e. when I come back one year later I can still
> understand why I wrote a specific part of the code. I can't say the same of the
> RecordParser, that's why I will not touch that code anymore - it just takes way
> too much time to understand what it does.

i never said that the record parser was good. that's why we wanted an antlr one in the first place. but still assembly is different from parsing

> Besides of the name conflicting thing, there is also a chicken-and-egg problem,
> see e.g.

chicken and egg has nothing to do with parsing. parsing is the activity of finding tokens and making simple objects from them. these objects don't have to be complete, we can fill them in at analysis. but that analysis is independent from how we parsed things. analysis can sit in the storage, or related to it, independent of the many parsers and the many grammar versions.

> at once and keep on adding fields because eventually the MDD will point to
> itself. In that case, the NamedResources will complain that the resource tries

this can be found at analysis. you can do it multi-pass.

at first pass you only know what MDDs exist (so you know whether ptr T or set T are valid because you know that T exists, you don't need to _parse_ T for that!)

at second pass, when you are done with this MDD (you know all its fields), you can start parsing all related ones, and so on. but you can even skip this until somebody actually asks you for T!

> to re-create itself, so the building needs to be dealt with in a separate step.

no it will not recreate itself because ptr Self only needs to know that Self exists, see above.

the confusion here starts from the fact that the first MDD parser was eager, so it was parsing all it was knowing about. this doesn't need to be the case, at least not in the first instance.

> For these reasons, and given the sheer amount of code that only deals with
> assembling the fields and making the sub-tables for sets etc. in
> DataDefinitionImpl, I want to move this code into a specific class. In the same

great, and that will be a parser-independent assembler, close to the storage. or?

but before _wanting_ to do things, i think it's good to have this discussion and then we will know what we all _want_.

> time, that would remove the knowledge of ASTs from DataDefinitionImpl and
> FieldDefinitionImpl.

yes, so you remove the connection between the parser and the assembler.

> > interface MDDStorage{
> >    void addField(String mddName, FieldDefinitionData d);
> >    void addRule(String mddName, RuleData r);
> >    void addFunction(String mddName, FunctionData f);
> > }
> 
> 
> given what I said above, that is not enough. I actually have no idea how you
> would model it in an interface...maybe

i still think it is enough. see the "spitting" above, and all my answers to your arguments.

note that the mddName can be "general.Person->languages", i.e. adding fields to subsets or subpointers. this also tells the storage (if it doesn't know already) that there _is_ a subset or a subpointer, and thus the storage can build the corresponding default fields. then when new (programmer) fields are added, collissions can be avoided.

> void addPointerRelToParentBeforePointerRelToForeignAndAvoidNameCollisions()
> will do.

PointerRelToParentBeforePointerRelToForeignAndAvoidNameCollisions is an _internal_ constraint of the storage... this is plain-old encapsulation! the parser doesn't have to know about it! it just sends a field.

> Seriously, the problem is not _that_ simple.

it is our job to make it simple through this discussion. when one programs something it's easy to overcomplicate it when one doesn't have a good analysis. now we have sufficient experience with this matter to make things simple.

> about the refactoring: my purpose with this entry was to make the new provider
> somewhat cleaner, given the concerns rudi raised in some mail (modularity with
> XML, ...), i.e. make the DataDefinitionImpl and FieldDefinitionImpl independent
> of AST, but _not_ to re-code the whole thing, introduce a MDDStorage, etc.,
> i.e. change the whole design of the provider.
> 
> so I think there are two stages:
> 
> 1) simple refactoring to make DataDefinitionImpl and FieldDefinitionImpl
> separate from AST, by introducing a factory that takes care of assembling MDDs,
> subfields etc.
> 2) complex refactoring, that involves much more work.
> 
> I am ok with doing level 1) but not more.

well my purpose here (and everywhere else) is to make makumba better. for that, I am OK with doing anything that is well thought through :) 

i seriously propose that you to use a corpus. that will make it easier to assess success and progress. for that you would need to _keep_ the present parser as it is (maybe moved in a test project/package), do a DD/FD comparator and make sure that the refactoring returns the same results as the previous code.

the alternative is to refactor into the existing one, but even then, if you do a comparator, you can compare to the RecordInfo storage on a large MDD corpus. validation definitions will fail but the crucial field ordering and name collissions will be caught.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 21 Apr 2010 10:13 UTC**
in order for the storage to figure out whether an MDD exists or not (see analysis above), an MDD finding strategy is needed. i'm not sure whether that needs to be parser dependent (in that case the parser must tell it to the storage). 

i think for now we can assume that finding the MDD is parser independent. 

the actual parsing handler/strategy can be decided based on file extension (once the mdd "source file" file is found by the strategy) and some comment in the file.

of course, a file finding strategy can be programmed _together_ with the parser, but they can work as separate entities.

also, more finding strategies could be configurable, and the priority order between them. 

but for now, once we made the assumption that finding and parsing are separate, we an use our current finding mechanism to figure out whether MDDs exist.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 22 Apr 2010 00:39 UTC**
a new version of MDDStorage:

interface MDDStorage{
   void addField(String mddName, FieldDefinitionData d);
   void addRule(String mddName, RuleData r);
   void addFunction(String mddName, FunctionData f);
   void finished(String mddName);
}

where finished() is called when the parser encounters the end of the respective MDD. this can signal to the storage that the MDD can be assembled, analyzed, cached, returned to clients, etc.

/* probably more methods are needed for directives like title. or title is a function now? !include should be dealt with by the parser */

this is similar to passing events to listeners. and *Data are data transfer objects :)

the advantage is that the parser doesn't have to keep an overview of the mdd, simply send things when it finds them. 

one aspect to be considered here is why not just have one method, which sends all the *Data at once, at finished time. one answer is that since the parser finds data from _different_ MDDs (the main one, the subsets, subpointers), it would need to keep field sets for more mdds, etc. even without that i still think this helps keep the parser simple.

designing the parser to be simple is crucial since there will be many parser types but most probably only one storage type.

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 22 Apr 2010 09:25 UTC**
(In reply to comment #7)
> one aspect to be considered here is why not just have one method, which sends
> all the *Data at once, at finished time. one answer is that since the parser
> finds data from _different_ MDDs (the main one, the subsets, subpointers), it
> would need to keep field sets for more mdds, etc. even without that i still
> think this helps keep the parser simple.

and here comes manu with the bad news: you _need_ to look into fields of other MDDs if you want to tell the user at MDD parse-time "look, your said here

!title = somePtr.someField

or

unique(somePtr.someField, someOtherField) : "These field have to be unique"

or some other case of the sort


but there is no field "someField" in type <type pointed by somePtr>".

Of course this works as well for ptr1.ptr2.ptr3.someField.


I'm open to suggestions here, it took me some efforts to be able to reproduce this (if I remember correctly there is a unit test from Stefan about this case, but if not we should add it). it happens very often that programmes mistype a field name, and then IMO, the problem should be displayed in the MDD they're currently working on, i.e. in the context.

------
about factory vs MDDStorage encapsulation: I like factory better because my gut feeling tells me that as soon as you'll want to go into more exotic things, like e.g. having an MDD provider for java beans (not so unrealistic if we want good JSF integration, will write about this when I have time), a factory gives you more power than an encapsulated object. I may be wrong tough, it would be a good idea to actually look what such factory would look like, if I find some time I will try to sketch one similar to the encapsulation sketch above.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 22 Apr 2010 10:54 UTC**
(In reply to comment #8)
> and here comes manu with the bad news: you _need_ to look into fields of other

again, the fact that you did this once does not entitle you to be always right :) actually quite the opposite, since you did it once and then agreeed that refactoring is needed

> !title = somePtr.someField
> unique(somePtr.someField, someOtherField) : "These field have to be unique"

no you don't. not during parsing. you can do this at a later pass of analysis, in the storage. parser-independent!

if you think of it, functions are a typical example of chicken-egg problem because they are in MDDs and they use MDDs. but in the end it's easy because they only need to know that the fields and functions they refer _exist_. you're not trying to see what type they have, if they can be analyzed now, etc. you do that type analysis later, after you are done with the _list_ of functions and fields from _this_ mdd.

title is a kind of function so it's part of this discussion.

So you have several levels of analysis
1) what mdds exist
2) what fields exist. what is the type of the fields (for pointer/set analysis, you use 1)
3) what functions exist, what constraints exist?
4) are constraints and functions correct? (for fields needed by functions and constraints, you use 2, for functions needed by functions, you use 3)

each of these levels use the levels above it. it does not try to do more and go into the levels below it.

by the time you start with level 4, parsing has ended. during 4, when you need another MDD (even the current one), you only need it up to step 3! 

if you want you can also ask it to do its step 4 just for it to be complete, but you don't _need_ that. in order to prevent re-parsing, you can use a stack to see how you got to the MDD you parse now, or simply a cache, where you put things immediately as you find them. this will avoid re-parsing because when you are at MDD B which uses the MDD A that we started from, we will find there the 1-3 of A, and use them

to be able to report errors properly, you simply include the line-column data and a reference to the text in all the *Data DTOs

so, this is not an issue. all compilers face this problem. that's why in C you have function declarations in header files, which ensures that a function is declared before it is called. to compile the code of a function, you only need to know that that functions it calls exist and it has a certain signature, not more than that.

> about factory vs MDDStorage encapsulation: I like factory better because my gut
> feeling tells me that as soon as you'll want to go into more exotic things,
> like e.g. having an MDD provider for java beans (not so unrealistic if we want
> good JSF integration, will write about this when I have time), a factory gives
> you more power than an encapsulated object. I may be wrong tough, it would be a
> good idea to actually look what such factory would look like, if I find some
> time I will try to sketch one similar to the encapsulation sketch above.

well since we clearly have issues with design and overdesign, i really suggest to do this in small steps, and with a corpus.

at first, you don't even need to have a proper getDataDefinition, with NamedResource caching and stuff. just produce DataDefinitions and compare them using the comparator to whatever the current parser or RecordInfo returns

at that step you will know that the parser-storage separation works, and that what you call "the hidden table contract" (only ordering is for tables, name collision avoidance is part of the storage internal logic) is OK. and you can think of MDD analysis, error printing, caching and so on. divide and conquer.

i think having DD and FD factories _per_ parser is fundamentally wrong. as we discussed, DD and FD assembly and storage is parser independent.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 22 Apr 2010 17:10 UTC**
to clarify a bit my proposal here, it is a simple separation between syntax (parser) and semantic (storage). the differences from the classical case are
- there's a many-to-one relationship between syntax and semantics. more parsers, one storage. that simply allows us to support more than one syntax and file format
- there's one storage for more "compilation units" (MDDs) but that's perfectly legitimate since they depend on each other. it's like the classloader in java.

one word about removing duplicate code. once you have a corpus, that's obsolete. the moment you know that the new parser works like the old and (where possible) like RecordInfo, you can safely throw away RecordInfo! or move it to some test project for future corpus tests.

about grammar, i don't really care, once the separation is done and parsers are kept to a minimum, it's easy to make another parser with a more complete grammar.

and i do believe that the current antlr parser code can be very easily separated into this structure.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 23 Apr 2010 00:44 UTC**
(In reply to comment #8)
>  you _need_ to look into fields of other
> MDDs if you want to tell the user at MDD parse-time "look, your said here

just a clarification: these are semantic errors. a parser only takes care of the syntax. these are not parse-time errors but analysis-time errors.

so no you don't tell such things to the user at parse-time. if the syntax is ok you make your simple objects and send them to the analyzer for chewing. you also return their file coordinates, just in case the analyzer finds an error and wants to refer back to the original source.

analysis time follows parse time so for the user it won't make any difference. but conceptually it's a big difference.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 23 Apr 2010 08:15 UTC**
and then there is the business of errors. to support nicely multiple syntactic errors (if we have a parser that can do that) we need to improve/overload the finished() method with something like

finished(String mdd, List<Exception> e)

where e is the list of syntactic errors found.

the storage can then decide whether to continue analyzing or just to store the syntactic errors, throw them, etc. to the list of syntactic errors, the storage can add the semantic errors it finds.

about syntactic errors, to what i see in the generated code, antlr-generated parsers can recover from errors and continue parsing to the end of the input, so when we get reportError() called, we can add the error to a list and let things flow. 

to what i could understand from the code (maybe i'm wrong), the current Mdd base parser only stores one error. that's also ok, that means we only report the first syntactic error. but in the general case a parser may find more syntactic errors, though many of them will be consequences of the first.

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 23 Apr 2010 08:34 UTC**
One other possible usage should be considered, namely the MDD viewer.
Currently, the viewer does its own, primitive parsing of some things of the MDD, to achieve syntax highlighting. This is erroneous, and is one more thing that needs to be kept in synch with the syntax.
Thus, the MDD viewer should rather use the parser result.

One possible idea would be that the MDD viewer implements the storage interface, and thus gets the parsed information easily from any type of parser.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 23 Apr 2010 09:20 UTC**
to continue on comment 13. (sorry this went to another bug, bugzilla usability sucks)

first, for source viewers to benefit (and i think they should, though maybe not in the first iteration), each DTO (FieldData, FunctionData) should not just be annotated with a pair of begin-end syntax points, but also syntax points for their internal syntactic components. for example

field= fixed ptr Type

will have syntax point annotations for field, fixed, ptr, Type

even semantic errors can benefit from these, since if above i find an error with the Type (e.g. the Type MDD doesn't exist) i can point directly to the type in my semantic error, not just generally to the field.

then architecturally, this can support viewers in two ways:

1) the viewer implements the MDDStorage (or we can call it MDDParserListener) contract. that means that it can react to whatever the parser finds

2) the viewer does not care about file type. the storage stores all these syntactic annotations, the viewer asks for them, finds also the file, and then shows the file and does blind, file-format-independent syntax coloring. this seems very powerful to me.

in both cases we assume that the viewer has access to the source text (Rudi used the example of a mysql show create table...)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 27 Apr 2010 08:24 UTC**
a word about management: i know that this bug started with different ambitions. i don't mean that we should have the architecture i propose right away, but we should _aim_ for it. this simply means that the bug will not close after the first refactoring iteration, yet that iteration can do what was originally planned, _keeping_in_mind_ (albeit not immediately adhering to) the wider architecture we agree upon (if we agree). but i don't think keeping the bug open a bit longer is an issue or?

----

so i would change the name of MDDStorage into MDDParserListener or so, since that's what it is. behind that interface, parsers can do what they want to provide the field/member data (including locations in source files for the member as a whole and for its constitutive elements), the errors about them, etc.

interface MDDParserListener{
   void addField(String mddName, FieldDefinitionData d);
   void addRule(String mddName, RuleData r);
   void addFunction(String mddName, FunctionData f);
   void finished(String mddName, List<MDDParseException> e);
}

due to this listener design, parsers do not need to keep any internal state about the MDD members they find. they simply find members and send them away.

in between parser and storage/analyzer, there can be a little class MddDTOAssembler that implements the above interface and assembles an MddDTO, with all the member data (*Data). i don't think it's a factory since factories make different things depending on their parameters. this simply puts smaller DTOs together in a larger DTO to deliver it all at once.

the MddDTO should be used both by the storage/assembler and can also be used by format-independent viewers. this way, the storage does not have to "wait" for all the MDD data to come in, but can get it all at once, as a collection of MddDTO

an MddDTOAssembler is finished (and can return what it collected) when all the MDDs it "hears" about are finished. e.g. when it parses general.Person it will also hear about general.Person->languages so it will make an MddDTO for that too. the job is finished when it finds out through finished() that general.Person->languages and general.Person are finished.

again, anything above the MDDParserListener interface (storage, MddDTO, MddDTOAssembler, even generic viewers) is parser-, file format- and grammar-independent.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 May 2010 10:11 UTC**
i thought a bit more of the FieldDefinitionData, RuleData and FunctionData, e.g. who would make them.

if they are really dumb DTOs, with a lot of data and not much functionality, this is not a big problem, the parser can make them and send them away.

however, it seems a bit natural to do some specialization, like different FieldData per field type, and maybe different RuleData for types of rules and different FunctionData for actors and other functions, static functions, etc.

if there is specialization, then it must be kept parser-independent, because in the end these things do not depend on syntax, though they may contain file-position data for error and source-view reasons.

so if we decide for specialized DTOs:
-  the storage should have factory methods to make such specialized *Data stuff, which the parser will just populate. 
- also if we have such things, then they could directly implement the *Definition interfaces (FieldDefinition, FunctionDefinition), so the storage doesn't have to make new objects when it's done analyzing them

so the (parser-independent) storage would get this parser-independent factory method:

FieldData makeFieldData(int type, SyntaxPoint fieldDefinition)

maybe this can be part of the parser listener interface, or a separate interface (as it's not really a listener method). The fieldDefinition syntax point would define the beginning and end of the field definition in the source file.

then the FieldData interface would have

setFixed(boolean)
setFixed(SyntaxPoint)
setUnique(boolean)
setUnique(SyntaxPoint)
setComment(String)
setComment(String, SyntaxPoint)

and so on. setFixed(boolean)  and other non-syntaxpoint methods would be invoked by the parsers that don't want to set a syntax point for the keyword that defines the fixed-ness.

and the charFieldData would have
setSize(int)
setSize(int, SyntaxPoint)

and the ptrFieldData would have
setPointedType(String)
setPointedType(String, SyntaxPoint)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 May 2010 13:43 UTC**
the design here is beautiful (besides, it's done by a modest designer) but i think we should drop it in favor of bug 1226 and bug 1229.

basically it does not make sense to extract DataDefinitions from more than the .mdd format, when you can directly make directly java classes from things like Mysql schema.  hibernate mappings already have the classes, you only need to add (really or virtually) some annotation.

actually, as per bug 1232, i would simply generate Class directly from parsing .mdd, no DD/FD any longer. of course we would generate FuncDef and ValidationDef.

so it seems to me that there's not much point in improving this really. there is repeated code, it was a nice seminar discussion on syntax/semantic separation but most of the code might go away by generating class directly, so why bother.

to generate class directly, we could open another bug (?) and maybe use some architectural elements discussed here.

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 24 May 2010 22:00 UTC**
(In reply to comment #17)
> the design here is beautiful (besides, it's done by a modest designer) but i
> think we should drop it in favor of bug 1226 and bug 1229.


One aspect that I was specifically interested in this design was the application to the (MDD) source code viewer. I didn't see that reflected in the 1226 and 1229, so one would need to do a source-specific viewer for each type of source...

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 25 May 2010 12:36 UTC**
(In reply to comment #18)
> One aspect that I was specifically interested in this design was the
> application to the (MDD) source code viewer. I didn't see that reflected in the
> 1226 and 1229, so one would need to do a source-specific viewer for each type
> of source...

well we will still need new parsers:
- for mdd if we translate directly to class
- for other formats that we don't have as class, like mysql schema

we do not need a parser for java/POJO though, as annotations are available already from the java compiler

also having some FieldData DTO containing just the data of FieldDefinition does make sense, also for mdd->class parser. so i will construct some FieldData (from FieldDefinition) when i make classes now. then maybe Manu wants to populate the FieldData with mdd file locations. 

then the parser can simply build FieldData and skip FieldDefinition, and use the FieldData to construct class (that code will already be available).

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:24 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
